blog: introducing juno (#2)
buildwithjuno/docs: blog/2023-02-14-introducing-juno/index.md

---
slug: introduction-juno
title: Introducing Juno 🚀
authors: [peterpeterparker]
tags: [hello world 👋]
---

I am thrilled to introduce [Juno](https://juno.build), a groundbreaking open-source Blockchain-as-a-Service solution that makes building Web3 applications more accessible to the millions of front-end developers worldwide.

![Overview screenshot](./overview.png)

---

## What is Juno?

Unlike traditional Backend-as-a-Service (BaaS) platforms like Google Firebase or AWS Amplify, Juno runs entirely on the blockchain.

> It provides you the ability to outsource all the behind-the-scenes aspects of your web or mobile development, giving you the ability to focus on the frontend.

Reviewer: ```suggestion
It allows you to forget about all the backend aspects of development, letting focus on what matters: the product your users see.
```

Just a suggestion, but in general I'm not sure "It provides you the ability" is grammatically correct

> At Juno, privacy is a top priority and the platform operates without controlling your data or work. With Juno, you truly own your creations.

Reviewer: This might be misinterpreted as "your data is private", which is a bit of a stretch if your data is replicated N times on privately owned replicas...

Author: Modified it to "Our platform operates without controlling your data or work. With Juno, you truly own your creations."

> The Juno closed beta program is launching today and is packed with all the essential features you need to start creating your Web3 applications.

Reviewer: ```suggestion
The Juno closed beta program is launching today and is packed with all the essential features you need to start creating Web3 applications.
```

or "your first Web3 application", but "your Web3 applications" sounds a little odd

---

> ## Concept

Reviewer: ```suggestion
## How It Works
```

or maybe "Concepts"

> Each project you start on Juno is a smart contract we called "Satellite" 🛰️.

Reviewer: ```suggestion
Every project you create on Juno is a "Satellite": a smart contract that you own, running 100% on-chain. 🛰️
```

> It is a container for your application that runs entirely on-chain and holds all the states or your project. It can contain your users' persisted data, a storage for the assets (images or video) and also your application bundle.

Reviewer: ```suggestion
It is a container for your application that holds all your project's state: users' persisted data, asset storage (images or video) and your application bundle.
```

(if you apply the suggestion above)

> A satellite lives on its own and you - and only you-can administrate it.

Reviewer: ```suggestion
A satellite lives on its own -- and only _you_ can administrate it, because you _own_ it, exactly like an NFT.
```

> In the future it will be enhanced with further capabilities and we might launch other extended canisters such as "Rocket" or "Space station".

Reviewer: I'm not super fond of making promises like this, since not only it points out to missing features but also makes it sound like (from your point of view) the project is not finished; also I'm not sure future capabilities are relevant to the user reading this right now

Author: It's my vision for the platform and it's also in the roadmap but, I have removed it to improve readability

We also plan to open the platform to third party developers allowing them to provide and sell their own templates.

---

## Authentication

Juno allows developers to securely identify users anonymously and save their data on the blockchain.

Our easy-to-use SDKs support authentication via [Internet Identity](https://internetcomputer.org/internet-identity), a Web3 authentication provider that offers a secure blockchain login experience with a user-friendly Web2 interface, and more providers will be added soon.

> Juno Authentication integrates tightly with other services, [datastore](https://juno.build/docs/build/datastore) and [storage](https://juno.build/docs/build/storage).

Reviewer: ```suggestion
Authentication integrates tightly with other Juno services, like [datastore](https://juno.build/docs/build/datastore) and [storage](https://juno.build/docs/build/storage):
```

```javascript
import { signIn } from '@junobuild/core';

const btn = document.querySelector("#signin");
btn?.addEventListener("click", signIn, { passive: true });
```

![Authentication screenshot](./authentication.png)

---

## Datastore

> Juno Datastore can be use to add persistency to your application. It uses a simple principle of collections and documents.

Reviewer: ```suggestion
Juno Datastore can be use to add persistence to your application using simple constructs: collections and documents.
```

> Each collections can handle which users and/or administrators have access to their data. Juno has never access to these and it is also worth to note that if you wish, you can also design your architecture in a way that even you the developer of your application can have no access to the data saved by your users.

Reviewer: ```suggestion
Collections specify which users and/or administrators have access to their data. Juno itself does _not_ have access to data in collections (unless you make it public, of course). Juno also gives you the possibility to design your architecture so that even _you_ (the developer of your application) can have no access to the data saved by your users.
```

You as the satellite's controller always have the possibility to change the code and remove the guards though, right?

Author: Correct. Until the day Juno offers a super easy way to transform a satellite or mission control into a sns.

```typescript
import { getDoc, setDoc } from '@junobuild/core';

const collection = "my_collection";
const key = "my_key";

const doc = await getDoc({
  collection,
  key
});

await setDoc({
  collection,
  doc: {
   key,
   ...(doc && doc),
   data: {
    email: myNewEmail
   }
  }
});
```

![Datastore screenshot](./datastore.png)

---

## Storage

> Juno Storage is the easiest way to let your users upload and manage their files securely on blockchain. The API we provide is designed to obfuscate any potential complexity. It takes advantage of the same read and write permission concept as the Datastore and automatically makes the file available on the Internet with the help of a built-in query parameters "token" mechanism that can be use to make users files secret.

Reviewer: ```suggestion
Juno Storage is the easiest way to let your users upload and manage their files securely on the blockchain. The API we provide is designed to obfuscate any potential complexity. It takes advantage of the same read and write permission concept as the Datastore and automatically makes the file available on the Internet with the help of a built-in query parameters "token" mechanism that can be use to make users files secret.
```

alternatively: on-chain

```javascript
const input = document.querySelector("input");

const {downloadUrl} = await uploadFile({
    data: input.files[0],
    collection: 'images'
});
```

![Storage screenshot](./storage.png)

---

## Hosting

> Juno Hosting - as for our other services - runs on the Internet Computer network. It is the only blockchain with canister smart contracts that can serve web and thus, without relying on any Big Tech infrastructure.

Reviewer: Wouldn't it make more sense to mention the IC earlier?

> Juno Hosting - as for our other services - runs on the Internet Computer network. It is the only blockchain with canister smart contracts that can serve web and thus, without relying on any Big Tech infrastructure.

Reviewer: ```suggestion
Juno Hosting - like our other services - runs on the Internet Computer. It is the only blockchain with canister smart contracts that can serve web assets to browsers straight from the chain -- meaning no dependency on centralized Big Tech.
```

> Therefore, the "Satellites" launch on Juno run entirely on blockchain, which allows for full decentralization.

Reviewer: ```suggestion
Therefore, the "Satellites" launched on Juno run entirely on blockchain, which allows for full decentralization.
```

maybe this is a little repetitive now?

> You can also maintain your unique brand identity with Juno Hosting as custom domain name are fully supported and can be configured with ease.

Reviewer: ```suggestion
Custom domain names are fully supported and easily configured, meaning your unique brand identity will be just as visible on chain.
```

![Hosting screenshot](./hosting.png)

---

## What's next?

> We are starting our closed beta program today. Fill this form - build with Juno - to enter the waiting list or contact David on [Twitter](https://daviddalbusco.com) (if you cannot wait to give Juno a try 😉).

Reviewer: ```suggestion
We are starting our closed beta program today. Fill this form - built with Juno - to enter the waiting list or contact David on [Twitter](https://daviddalbusco.com) (if you cannot wait to give Juno a try 😉).
```

> GitHub Star are also warmly welcomed. Drop a ⭐️ on our [GitHub repo](https://github.com/buildwithjuno/juno).

Reviewer: ```suggestion
⭐️⭐️⭐️ are also much appreciated: visit the [GitHub repo](https://github.com/buildwithjuno/juno) and show your support!
```
